### PR TITLE
Add listener functions to views

### PR DESCRIPTION
**Pull request by vegaro**
Depends on https://github.com/RevenueCat/purchases-hybrid-common/pull/706

TODO:

- [x] Android FooterPaywallView
- [x] iOS FooterPaywallView
- [x] Android PaywallView
- [x] iOS PaywallView

### CODE REVIEW

File: react-native-purchases-ui/ios/RCPaywallFooterViewManager.m [hunk 1]
```diff
@@ -65,7 +73,39 @@ - (void)safeAreaInsetsDidChange {
 #pragma clang diagnostic pop
 }
 
-- (void)paywallViewController:(RCPaywallViewController *)controller didChangeSizeTo:(CGSize)size API_AVAILABLE(ios(15.0)){
+- (void)                paywallViewController:(RCPaywallViewController *)controller
```
Comment (vegaro) [inline react-native-purchases-ui/ios/RCPaywallFooterViewManager.m:76]: This needs https://github.com/RevenueCat/purchases-hybrid-common/pull/706

File: react-native-purchases-ui/src/index.tsx [hunk 2]
```diff
@@ -77,12 +82,38 @@ type FullScreenPaywallViewProps = {
   style?: StyleProp<ViewStyle>;
   children?: ReactNode;
   options?: FullScreenPaywallViewOptions;
+  // Only Android
+  onPurchaseStarted?: (aPackage: PurchasesPackage) => void;
```
Comment (vegaro) [inline react-native-purchases-ui/src/index.tsx:86]: This requires an update in iOS PHC

**Comment by vegaro** [comment]
@RevenueCat/coresdk I still need to do anther pass of testing in iOS, but I would appreciate a first pass on reviews

File: react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallViewManager.kt [hunk 3]
```diff
@@ -19,7 +17,10 @@ internal class PaywallViewManager : BasePaywallViewManager<PaywallView>() {
     }
 
     override fun createViewInstance(themedReactContext: ThemedReactContext): PaywallView {
-        return PaywallView(themedReactContext)
+        return PaywallView(themedReactContext).also {
+            it.setPaywallListener(createPaywallListenerWrapper(themedReactContext, it))
+            it.setDismissHandler(getDismissHandler(themedReactContext, it))
```
Comment (tonidero) [inline react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallViewManager.kt:22]: Same here about passing these in the constructor

**Review by tonidero — COMMENTED** [review]
Left some comments but nothing blocking. Lmk when this is ready for a final pass!

File: examples/purchaseTesterTypescript/app/screens/FooterPaywallScreen.tsx [hunk 4]
```diff
@@ -4,15 +4,64 @@ import RevenueCatUI from 'react-native-purchases-ui';
...
+  const onRestoreCompleted = (customerInfo: CustomerInfo) => {
+    console.log('Restore completed:', customerInfo);
+  };
+
+  const onRestoreError = (error: PurchasesError) => {
+    console.error('Restore error:', error);
+  };
+
+  const onDismiss = () => {
+    console.error('Dismissed');
```
[the remark this answers is not in the document]
Comment (vegaro) [reply examples/purchaseTesterTypescript/app/screens/FooterPaywallScreen.tsx:49]: yes, good catch, that was some copy pasta :D

File: react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallFooterViewManager.kt [hunk 5]
```diff
@@ -59,6 +56,9 @@ internal class PaywallFooterViewManager : BasePaywallViewManager<PaywallFooterVi
                     }
                 }
             }
+        }.also {
+            it.setPaywallListener(createPaywallListenerWrapper(themedReactContext, it))
+            it.setDismissHandler(getDismissHandler(themedReactContext, it))
```
[the remark this answers is not in the document]
Comment (vegaro) [reply react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallFooterViewManager.kt:61]: Oh I think I've missed that

[the remark this answers is not in the document]
Comment (vegaro) [reply react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallFooterViewManager.kt:61]: hmm, the view id is required for sending the events in the both the paywall listener and the dismiss handler. So I think we need to create the view first sadly

File: react-native-purchases-ui/src/index.tsx [hunk 6]
```diff
@@ -77,12 +82,38 @@ type FullScreenPaywallViewProps = {
...
+                           customerInfo,
+                           storeTransaction
+                         } : {customerInfo: CustomerInfo, storeTransaction: PurchasesStoreTransaction}) => void;
+  onPurchaseError?: (error: PurchasesError) => void;
+  onPurchaseCancelled?: () => void;
+  // Only Android
+  onRestoreStarted?: () => void;
+  onRestoreCompleted?: (customerInfo: CustomerInfo) => void;
+  onRestoreError?: (error: PurchasesError) => void;
+  onDismiss?: () => void;
```
[the remark this answers is not in the document]
Comment (vegaro) [reply react-native-purchases-ui/src/index.tsx:97]: hmm they are not really options, they are events. I don't know if in Javascript that's cleaner tbh. I will ask someone from the frontend team

File: react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallFooterViewManager.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/PaywallFooterViewManager.kt:61]: Ohhh ok... That's ok then.

**Comment by vegaro** [comment]
@NachoSoto I have been testing iOS for a while, but I am not sure why the `delegate` is nil when it hits these lines:

https://github.com/RevenueCat/purchases-ios/blob/b99c8a6f6392038f26c68240a87c50519f053316/RevenueCatUI/UIKit/PaywallViewController.swift#L254

Is it maybe because `delegate` is weak?

**Comment by vegaro** [comment]
this was an issue https://github.com/RevenueCat/purchases-hybrid-common/pull/724 but even after that fix, the delegate is null

**Comment by NachoSoto** [comment]
@vegaro yes, if `PaywallViewController`'s delegate (`PaywallViewController`, since it also uses `[weak self]` for those closures) aren't held in memory, then it'll become `nil`.

File: react-native-purchases-ui/ios/PaywallViewWrapper.m [hunk 7]
```diff
@@ -73,4 +73,37 @@ - (void)setOptions:(NSDictionary *)options {
     }
 }
 
+- (void)                paywallViewController:(RCPaywallViewController *)controller
```
Comment (NachoSoto) [inline react-native-purchases-ui/ios/PaywallViewWrapper.m:76]: Nit: you have a lot of space here?

File: react-native-purchases-ui/src/index.tsx [hunk 8]
```diff
@@ -77,12 +82,30 @@ type FullScreenPaywallViewProps = {
   style?: StyleProp<ViewStyle>;
   children?: ReactNode;
   options?: FullScreenPaywallViewOptions;
+  onPurchaseCompleted?: ({
+                           customerInfo,
+                           storeTransaction
+                         }: { customerInfo: CustomerInfo, storeTransaction: PurchasesStoreTransaction }) => void;
```
Comment (tonidero) [inline react-native-purchases-ui/src/index.tsx:88]: Hmm I wonder if we should create types for these... But it's probably ok for now.

**Review by tonidero — APPROVED** [review]
Looks great! 🎉

**Pull request merged**